src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Automation/UiaTextRange.cs

```diff
@@ -0,0 +1,1093 @@
+                {
+                    return 0;
+                }
+
+                return moved;
+            }
+
+            if (count < 0)
+            {
+                if (moveStart)
+                {
+                    Start = MoveEndpointBackward(Start, unit, count, out moved);
+
+                    // If the start did not change then no move was done.
+                    if (start == Start)
+                    {
+                        return 0;
+                    }
+
+                    return moved;
+                }
+
+                End = MoveEndpointBackward(End, unit, count, out moved);
+
+                // If the end did not change then no move was done.
+                if (end == End)
+                {
+                    return 0;
+                }
+
+                return moved;
+            }
+
+            // Moving zero of any unit has no effect.
+            return 0;
+        }
+
+        void ITextRangeProvider.MoveEndpointByRange(
+            TextPatternRangeEndpoint endpoint,
+            ITextRangeProvider targetRange,
+            TextPatternRangeEndpoint targetEndpoint)
+        {
+            UiaTextRange textRange = (UiaTextRange)targetRange;
+            int e = (targetEndpoint == TextPatternRangeEndpoint.Start)
+                ? textRange.Start
+                : textRange.End;
+
+            if (endpoint == TextPatternRangeEndpoint.Start)
+            {
+                Start = e;
+            }
+            else
+            {
+                End = e;
+            }
+        }
+
+        void ITextRangeProvider.Select()
+        {
+            _provider.SetSelection(Start, End);
+        }
+
+        void ITextRangeProvider.AddToSelection()
+        {
+            // Do nothing. Do not throw exception.
+        }
+
+        void ITextRangeProvider.RemoveFromSelection()
+        {
+            // Do nothing. Do not throw exception.
+        }
+
+        void ITextRangeProvider.ScrollIntoView(BOOL alignToTop)
+        {
+            if (_provider.IsMultiline)
+            {
+                int newFirstLine;
+
+                if (alignToTop.IsTrue())
+                {
+                    newFirstLine = _provider.GetLineFromCharIndex(Start);
+                }
+                else
+                {
+                    newFirstLine = Math.Max(0, _provider.GetLineFromCharIndex(End) - _provider.LinesPerPage + 1);
+                }
+
+                _provider.LineScroll(Start, newFirstLine - _provider.FirstVisibleLine);
+
+                return;
+            }
+
+            if (_provider.IsScrollable)
+            {
+                _provider.GetVisibleRangePoints(out int visibleStart, out int visibleEnd);
+
+                short key;
+
+                if (_provider.IsReadingRTL)
+                {
+                    key = VK.LEFT;
+
+                    if (Start > visibleStart)
+                    {
+                        key = VK.RIGHT;
+                    }
```

Comment:
I suggest to move it outside of condition thread and replace by:
```suggestion
                    short key = Start > visibleStart ? VK.RIGHT : VK.LEFT;
```